pkg/diagnostics/client/run_diagnostics_pod.go

```diff
@@ -0,0 +1,163 @@
+func (d *DiagnosticPod) Name() string {
+	return DiagnosticPodName
+}
+
+// Description is part of the Diagnostic interface and provides a user-focused description of what the diagnostic does.
+func (d *DiagnosticPod) Description() string {
+	return "Create a pod to run diagnostics from the application standpoint"
+}
+
+// CanRun is part of the Diagnostic interface; it determines if the conditions are right to run this diagnostic.
+func (d *DiagnosticPod) CanRun() (bool, error) {
+	if d.PreventModification {
+		return false, fmt.Errorf("Running the diagnostic pod is an API change, which is prevented as you indicated.")
+	} else {
+		return true, nil
+	}
+}
+
+// Check is part of the Diagnostic interface; it runs the actual diagnostic logic
+func (d *DiagnosticPod) Check() types.DiagnosticResult {
+	r := types.NewDiagnosticResult("DiagnosticPod")
+	d.runDiagnosticPod(nil, r)
+	return r
+}
+
+func (d *DiagnosticPod) runDiagnosticPod(service *kapi.Service, r types.DiagnosticResult) {
+	loglevel := d.Level
+	if loglevel > 2 {
+		loglevel = 2 // need to show summary at least
+	}
+	imageName := d.ImageTemplate.ExpandOrDie("deployer")
+	pod, err := d.KubeClient.Pods(d.Namespace).Create(&kapi.Pod{
+		ObjectMeta: kapi.ObjectMeta{GenerateName: "pod-diagnostic-test-"},
+		Spec: kapi.PodSpec{
+			RestartPolicy: kapi.RestartPolicyNever,
+			Containers: []kapi.Container{
+				{
+					Name:    "pod-diagnostics",
+					Image:   imageName,
+					Command: []string{"pod-diagnostics", "-l", strconv.Itoa(loglevel)},
+				},
+			},
+		},
+	})
+	if err != nil {
+		r.Error("DCli2001", err, fmt.Sprintf("Creating diagnostic pod with image %s failed. Error: (%[2]T) %[2]v", imageName, err))
+		return
+	}
+	defer func() { // delete what we created, or notify that we couldn't
+		zero := int64(0)
+		delOpts := kapi.DeleteOptions{TypeMeta: pod.TypeMeta, GracePeriodSeconds: &zero}
+		if err := d.KubeClient.Pods(d.Namespace).Delete(pod.ObjectMeta.Name, &delOpts); err != nil {
+			r.Error("DCl2002", err, fmt.Sprintf("Deleting diagnostic pod '%s' failed. Error: %s", pod.ObjectMeta.Name, fmt.Sprintf("(%T) %[1]s", err)))
+		}
+	}()
+	pod, err = d.KubeClient.Pods(d.Namespace).Get(pod.ObjectMeta.Name) // status is filled in post-create
+	if err != nil {
+		r.Error("DCli2003", err, fmt.Sprintf("Retrieving the diagnostic pod definition failed. Error: (%T) %[1]v", err))
+		return
+	}
+	r.Debug("DCli2004", fmt.Sprintf("Created diagnostic pod named %v running image %s.", pod.ObjectMeta.Name, imageName))
+
+	bytelim := int64(1024000)
+	podLogsOpts := &kapi.PodLogOptions{
+		TypeMeta:   pod.TypeMeta,
+		Container:  "pod-diagnostics",
+		Follow:     true,
+		LimitBytes: &bytelim,
+	}
+	req, err := d.Factory.LogsForObject(pod, podLogsOpts)
+	if err != nil {
+		r.Error("DCli2005", err, fmt.Sprintf("The request for diagnostic pod logs failed unexpectedly. Error: (%T) %[1]v", err))
+		return
+	}
+
+	// wait for pod to be started and logs available
+	var scanner *bufio.Scanner
```

Comment:
If I look at the caveats under https://golang.org/pkg/bufio/#Scanner it seems to me this would not really be a concern in the case where the text being scanned is well-behaved. We control the log output of this pod, so I don't see this as a problem here, and count myself more likely to introduce bugs by going to a lower-level bufio.Reader implementation.